api/envoy/config/common/matcher/v3/matcher.proto

```diff
@@ -142,6 +143,15 @@ message Matcher {
   // to have matched.
   // If not specified, the matcher is considered not to have matched.
   OnMatch on_no_match = 3;
+
+  // [#not-implemented-hide:]
+  // When the matching condition is met, ``skip_percentage`` controls the
+  // percentage of traffic that should be skipped, i.e, not going through
+  // the matching action. By default this is 0.
+  // This is to support the sampling behavior. For example, if only 30% requests
+  // need to be sent to the logging server, then one can config the
+  // ``skip_percentage`` as 1 - 30% = 70%.
+  type.v3.FractionalPercent skip_percentage = 4;
```

Comment:
QQ: why is this not a probabilistic matcher condition itself that returns true/false with 30% probability, say?

Reply:
+1.  I don't think we need a new top-level field for this.  We should be able to plug this into the matcher framework as we would any other matcher.

Reply:
@kyessenov this field configs the sampling ratio, i.e, 30%.  Then in the implementation, we will have a random generator to generate a true/false condition (coin toss) based on this ratio.  

Also, we want to have this sampling executed after matcher condition met. The idea is that: say if 30% traffic enter ext_proc filter will be sent to side stream server. That's 30% traffic after the xDS matcher->ext_proc.    It's not the 30% total traffic go in this xDS matcher.

Reply:
@markdroth Another place I had been thinking to add this configuration is here: https://github.com/envoyproxy/envoy/blob/838bc86a0fe46801320eef13cc599bc80bd88d10/api/envoy/extensions/common/matching/v3/extension_matcher.proto#L25
and the PerRoute config:
https://github.com/envoyproxy/envoy/blob/838bc86a0fe46801320eef13cc599bc80bd88d10/api/envoy/extensions/common/matching/v3/extension_matcher.proto#L25

And the sampling logic will be executed in the match_delegate filter: https://github.com/envoyproxy/envoy/blob/838bc86a0fe46801320eef13cc599bc80bd88d10/source/extensions/filters/http/match_delegate/config.cc#L149

I am working on a design doc on this.

Reply:
I think the suggestion is to just make this an input matcher, but not matching on any actual input (just doing the dice roll). I.e. something that plugs in https://github.com/envoyproxy/envoy/blob/838bc86a0fe46801320eef13cc599bc80bd88d10/api/envoy/config/common/matcher/v3/matcher.proto#L66. 

The strongest counter-argument is that this isn't actually an input being matched on, but I don't feel that strongly.

Reply:
I think it would be easier to compose with a dice roll input. For example, we can make it stateful - e.g. return once every 1000 rolls consistently, and then express something like "report access log once every 1000 requests for 200, but always for != 200". We should definitely document it either way with an example, since matchers in general are pretty hard to write by hand.

Reply:
That's kind of cool actually, since we could also source dice roll from headers as needed etc.

Reply:
why cant we simply call this as sampling percentage instead of calling it as skip percentage?

Reply:
@rshriram , If we add a sampling percentage field, then by default this field has to be 100%, which does not comply with the proto default value, which is zero.  That's the reason we added it the opposite way.

Reply:
Moving the API change to action phase under ExecuteFilterAction.  This makes the dice roll on specific targets which matched the condition.

Reply:
The complete change to support this idea is in: https://github.com/envoyproxy/envoy/pull/34611